### PR TITLE
warn on missing dots for net5.0+ TFMs

### LINKED ISSUE #9215: NuGet pack should catch incorrect omitting of the dot in version numbers
For .NET 5, [we're going](https://github.com/dotnet/designs/pull/92) to reuse the existing `net` moniker. Due to our new versioning scheme (fixed schedule, major bump every year) we're only five years away from .NET 10. Developers might accidentally say `net10` when they really meant `net10.0`. We're planning to follow in the footsteps of .NET Core where project files will generally use a dot, even when it's not necessary (e.g. `net5.0`) but of course this won't prevent humans from dropping the `.0` either inadvertently or for aesthetics.

It was [suggested](https://github.com/dotnet/designs/pull/92#discussion_r383463597) to handle this as part of NuGet pack.

There are several ways to deal with that:

1. Generally warn when a dot is missing
2. Only warn on when the TFM is >= `net5.0` (regardless of whether there is an OS flavor or not)

The (1) is more correct but will cause a lot of noise, especially in .NET Framework projects where the ambiguity is irrelevant as the platform will remain at 4.x forever. So my preference would be something more targeted. Not sure (2) is the best way to address that but it seems to hit the sweet spot.

### PR DESCRIPTION
**Pull request by zkat**
## Bug

Fixes: https://github.com/nuget/home/issues/9215
Regression:No  
* How are we preventing it in future:   tests added

## Fix

This makes it so we warn when a user uses something like `net50` instead of `net5.0`. We want customers to start getting used to (and will eventually enforce) including dots in .NET version numbers.

## Testing/Validation

Tests Added: Yes

### CODE REVIEW

File: src/NuGet.Core/NuGet.Build.Tasks.Pack/PackTaskLogic.cs [hunk 1]
```diff
@@ -57,9 +57,40 @@ public PackArgs GetPackArgs(IPackTaskRequest<IMSBuildItem> request)
...
             var aliases = new Dictionary<string, string>();
-            foreach (var tfm in assetsFile.PackageSpec.TargetFrameworks)
+            foreach (var targetFramework in assetsFile.PackageSpec.TargetFrameworks)
             {
-                aliases[tfm.TargetAlias] = tfm.FrameworkName.GetShortFolderName();
+                string alias = targetFramework.TargetAlias;
+                NuGetFramework framework = targetFramework.FrameworkName;
+                aliases[alias] = framework.GetShortFolderName();
+
+                var isNet5EraTfm = framework.Version.Major >= 5 && StringComparer.OrdinalIgnoreCase.Equals(FrameworkConstants.FrameworkIdentifiers.NetCoreApp, framework.Framework);
```
Comment (zivkan) [inline src/NuGet.Core/NuGet.Build.Tasks.Pack/PackTaskLogic.cs]: I think it was in the previous PR I came up with a name something like "tfmWithPlatform".

On one hand isNet5Era is used frequently throughout the code, so it's consistent using it again here. On the other name it's a horrible name that won't make sense in a few months when all development is focused on net6 and higher.

File: src/NuGet.Core/NuGet.Build.Tasks.Pack/PackTaskLogic.cs [hunk 2]
```diff
@@ -57,9 +57,40 @@ public PackArgs GetPackArgs(IPackTaskRequest<IMSBuildItem> request)
...
+                if (isNet5EraTfm)
+                {
+                    try
+                    {
+                        var parsedAlias = NuGetFramework.Parse(alias);
```
Comment (zivkan) [inline src/NuGet.Core/NuGet.Build.Tasks.Pack/PackTaskLogic.cs]: Can you remind me why we're doing this check in `PackTask`?

My memory is that we agreed that the SDK should warn on the same scenario on restore (or evaluation). Restore creates the assets file, and this pack task reads the assets file. Therefore, to get to this code, the customer must already have got the warning from the SDK (once it's implemented).

File: src/NuGet.Core/NuGet.Packaging/Core/FrameworkNameValidatorUtility.cs [hunk 3]
```diff
@@ -34,6 +38,70 @@ internal static bool IsValidFrameworkName(string path)
...
+        internal static bool IsDottedFrameworkVersion(string path)
+        {
+            foreach (string knownFolder in PackagingConstants.Folders.SupportFrameworks)
+            {
+                string folderPrefix = knownFolder + System.IO.Path.DirectorySeparatorChar;
+                if (path.Length > folderPrefix.Length &&
+                    path.StartsWith(folderPrefix, StringComparison.OrdinalIgnoreCase))
+                {
+                    var managedCodeConventions = new ManagedCodeConventions(new RuntimeGraph());
+                    var collection = new ContentItemCollection();
```
Comment (zivkan) [inline src/NuGet.Core/NuGet.Packaging/Core/FrameworkNameValidatorUtility.cs]: Although pack isn't the most performance critical code path, creating a new instance of `ManageCodeConventions` and `ContentItemCollection` per file being checked doesn't seem very good.

I would suggest:
* Create one ManagedCodeConventions instance and one ContentItemCollection instance in InvalidUndottedFrameworkRule
* Add all the files in the package into the collection.
* Iterate the list of PatternSets that contain TFM (RuntimeAssemblies, CompileRefAssemblies, CompileLibAssemblies, etc) and call `ContentExtractor`.

Unless you have some benchmarks showing that pre-filtering files via your `PackagingConstances.FOlders.SupportFrameworks` and `path.StartsWith` is worthwhile, I think it's just duplicating work that the managed code conventions and content extractor will already do.

I would also avoid using `managedCodeConventions.Patterns.AnyTargettedFile` because it doesn't match the runtimes file pattern (you've only put in an exception for `contentFiles`), and if you take my advise about not using `path.StartsWith` to pre-filter the folders, then we don't want `notused/net50` to generate the warning, because NuGet will never select assets from a folder named `notused`.

File: src/NuGet.Core/NuGet.Packaging/Core/FrameworkNameValidatorUtility.cs [hunk 4]
```diff
@@ -34,6 +38,70 @@ internal static bool IsValidFrameworkName(string path)
             return fx != null && fx.Framework != NuGetFramework.UnsupportedFramework.Framework;
         }
 
+        internal static bool IsDottedFrameworkVersion(string path)
```
Comment (zivkan) [inline src/NuGet.Core/NuGet.Packaging/Core/FrameworkNameValidatorUtility.cs]: I'd suggest making this method private inside the validation rule.

Firstly, the method name is `IsDottedFrameworkVersion`, but the input is not a framework string. The method is extracting the TFM from a path, and then doing what the name suggests, so it fails the "single responsibility principal", plus the naming convention where a method name should describe what it's doing.

Secondly, it returns "true" when the path doesn't contain a TFM. That's a pretty significant gotcha in case anyone wants to re-use this method in the future.

File: src/NuGet.Core/NuGet.Build.Tasks.Pack/PackTaskLogic.cs (hunk 2 above)
Comment (zkat) [reply src/NuGet.Core/NuGet.Build.Tasks.Pack/PackTaskLogic.cs]: I thought we were doing this in case they packed a long time after doing a restore?

Comment (nkolev92) [reply src/NuGet.Core/NuGet.Build.Tasks.Pack/PackTaskLogic.cs]: I think the SDK should be warning here not NuGet.

File: src/NuGet.Core/NuGet.Common/Errors/NuGetLogCode.cs [hunk 5]
```diff
@@ -857,5 +857,10 @@ public enum NuGetLogCode
         /// Undefined package warning
         /// </summary>
         NU5500 = 5500,
+
+        /// <summary>
+        /// InvalidUndottedFrameworkWarning
+        /// </summary>
+        NU5501 = 5501,
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Common/Errors/NuGetLogCode.cs:864]: We'll probably need a docs issue to add docs for this warning.

File: src/NuGet.Core/NuGet.Packaging/PackagingConstants.cs [hunk 6]
```diff
@@ -1,6 +1,7 @@
 // Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
+using System.Collections.Generic;
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Packaging/PackagingConstants.cs]: Unnecessary change?

File: src/NuGet.Core/NuGet.Packaging/Rules/InvalidUndottedFrameworkRule.cs [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+            var set = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+            foreach (var file in builder.GetFiles().Select(t => PathUtility.GetPathWithDirectorySeparator(t)))
+            {
+                set.Add(file);
+            }
+
+            var managedCodeConventions = new ManagedCodeConventions(new RuntimeGraph());
+            var collection = new ContentItemCollection();
+            collection.Load(set.Select(path => path.Replace('\\', '/')).ToArray());
+
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Packaging/Rules/InvalidUndottedFrameworkRule.cs:205]: format: extra line.

File: src/NuGet.Core/NuGet.Packaging/Rules/InvalidUndottedFrameworkRule.cs [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+
+            var patterns = managedCodeConventions.Patterns;
+
+            var frameworkPatterns = new List<PatternSet>()
+            {
+                patterns.RuntimeAssemblies,
+                patterns.CompileRefAssemblies,
+                patterns.CompileLibAssemblies,
+                patterns.NativeLibraries,
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Packaging/Rules/InvalidUndottedFrameworkRule.cs:213]: Does native have a `{tfm}` in its pattern?

File: src/NuGet.Core/NuGet.Packaging/Rules/InvalidUndottedFrameworkRule.cs [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+
+                            if (framework.Version.Major >= 5 &&
+                                StringComparer.OrdinalIgnoreCase.Equals(FrameworkConstants.FrameworkIdentifiers.NetCoreApp, framework.Framework))
+                            {
+                                var dotIndex = frameworkString.IndexOf('.');
+                                var dashIndex = frameworkString.IndexOf('-');
+                                var frameworkVersionHasDots = (dashIndex > -1 && dotIndex > -1 && dotIndex < dashIndex) || (dashIndex == -1 && dotIndex > -1);
+                                if (!frameworkVersionHasDots)
+                                {
+                                    messages.Add(CreatePackageIssue(item.Path));
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Packaging/Rules/InvalidUndottedFrameworkRule.cs]: Can we create one warning and just list the offending file paths? 

We have followed `1 vs n` warning approach in the past with: https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Packaging/Rules/DependenciesGroupsForEachTFMRule.cs. 

Restore also merges the same code warnings on the same target graph (same TFM/RID combo)

File: src/NuGet.Core/NuGet.Packaging/Rules/InvalidUndottedFrameworkRule.cs [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+                                }
+                            }
+                        }
+                        catch (ArgumentException)
+                        {
+                            // if the parsing fails, we treat it as if this file
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Packaging/Rules/InvalidUndottedFrameworkRule.cs]: Can this happen with the new approach?

File: src/NuGet.Core/NuGet.Build.Tasks.Pack/PackTaskLogic.cs (hunk 2 above)
Comment (zkat) [reply src/NuGet.Core/NuGet.Build.Tasks.Pack/PackTaskLogic.cs]: cc @wli3 is there an issue on your end to track this? Should we create one for you?

@nkolev92 @zivkan When you say we shouldn't do this check in `PackTask`, do you mean I should _also_ not add the rule? Like, to be clear, are you saying this entire PR is unnecessary?

Comment (nkolev92) [reply src/NuGet.Core/NuGet.Build.Tasks.Pack/PackTaskLogic.cs]: I think we're referring to this particular check. 
This is validating the `TargetFramework` in the csproj property value. NuGet shouldn't do that.
We should validate the missing dots for anything the customers might have created manually for one or another reason (such as validating the nuspec + files in the package).

Comment (zkat) [reply src/NuGet.Core/NuGet.Build.Tasks.Pack/PackTaskLogic.cs]: Alright. So we'll have to make sure the SDK agrees with the assessment and takes the time to implement this on their end before we can say we're complete. Would it make sense to do it on our end _now_, since it's implemented, and our timeline is so short, and switch to an SDK-sourced error later? /cc @aortiz-msft

Comment (nkolev92) [reply src/NuGet.Core/NuGet.Build.Tasks.Pack/PackTaskLogic.cs]: I really don't think NuGet should get in the business of raising a warning for the TargetFramework property even temporarily. We spent lots of time getting away from it :) 

The way I see it, that specific warning is a nice to have. 

* The warning for the `TargetFramework` property only impacts the project author. 
* The warning on the package assets can affect all consumers of that package. 

I'd argue the latter group is larger.

Comment (wli3) [reply src/NuGet.Core/NuGet.Build.Tasks.Pack/PackTaskLogic.cs]: Do I still need to do anything on my side?

Comment (zkat) [reply src/NuGet.Core/NuGet.Build.Tasks.Pack/PackTaskLogic.cs]: @nkolev92 wait, do the validation rules run when you install/restore, too?...

Comment (zkat) [reply src/NuGet.Core/NuGet.Build.Tasks.Pack/PackTaskLogic.cs]: @wli3 most likely, yes. You'd have to warn when, for example, someone does `<TargetFramework>net50</TargetFramework>` (that is, when you're resolving an alias that looks like a concrete targetframework).

File: src/NuGet.Core/NuGet.Packaging/Rules/InvalidUndottedFrameworkRule.cs (hunk 8 above)
Comment (zkat) [reply src/NuGet.Core/NuGet.Packaging/Rules/InvalidUndottedFrameworkRule.cs:213]:
```csharp
                NativeLibraries = new PatternSet(
                    conventions.Properties,
                    groupPatterns: new PatternDefinition[]
                        {
                            new PatternDefinition("runtimes/{rid}/nativeassets/{tfm}/{any?}", table: DotnetAnyTable),
                            new PatternDefinition("runtimes/{rid}/native/{any?}", table: null, defaults: DefaultTfmAny)
                        },
                    pathPatterns: new PatternDefinition[]
                    {
                        new PatternDefinition("runtimes/{rid}/nativeassets/{tfm}/{any}", table: DotnetAnyTable),
                        new PatternDefinition("runtimes/{rid}/native/{any}", table: null, defaults: DefaultTfmAny)
                    });
```

yepp

File: src/NuGet.Core/NuGet.Common/Errors/NuGetLogCode.cs (hunk 5 above)
Comment (zkat) [reply src/NuGet.Core/NuGet.Common/Errors/NuGetLogCode.cs:864]: https://github.com/NuGet/docs.microsoft.com-nuget/issues/2177

**Comment by zkat** [comment]
PR for docs page: https://github.com/NuGet/docs.microsoft.com-nuget/pull/2178

File: src/NuGet.Core/NuGet.Packaging/Rules/InvalidUndottedFrameworkRule.cs [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+        {
+            var set = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+            foreach (var file in builder.GetFiles().Select(t => PathUtility.GetPathWithDirectorySeparator(t)))
+            {
+                set.Add(file);
+            }
+
+            var managedCodeConventions = new ManagedCodeConventions(new RuntimeGraph());
+            var collection = new ContentItemCollection();
+            collection.Load(set.Select(path => path.Replace('\\', '/')).ToArray());
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Packaging/Rules/InvalidUndottedFrameworkRule.cs:204]: hmm...this seems unusual, are we doing this in other rules as well?

I'd have expected the build to normalize them already.

File: src/NuGet.Core/NuGet.Packaging/Rules/InvalidUndottedFrameworkRule.cs [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+                        var framework = (NuGetFramework)item.Properties["tfm"];
+                        if (framework == null)
+                        {
+                            continue;
+                        }
+
+                        if (framework.Version.Major >= 5 &&
+                            StringComparer.OrdinalIgnoreCase.Equals(FrameworkConstants.FrameworkIdentifiers.NetCoreApp, framework.Framework))
+                        {
+                            var dotIndex = frameworkString.IndexOf('.');
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Packaging/Rules/InvalidUndottedFrameworkRule.cs]: This is a great candidate for a unit test, I'd feel better if we had a bunch of unit test for a utility that just checks the dots in the string.

File: src/NuGet.Core/NuGet.Packaging/Rules/InvalidUndottedFrameworkRule.cs [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+        public string MessageFormat { get; }
+
+        public InvalidUndottedFrameworkRule(string messageFormat)
+        {
+            MessageFormat = messageFormat;
+        }
+
+        public IEnumerable<PackagingLogMessage> Validate(PackageArchiveReader builder)
+        {
+            var set = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Packaging/Rules/InvalidUndottedFrameworkRule.cs:196]: Can we validate the potential framework values in the nuspec which could be hand edited. 
See reference: https://docs.microsoft.com/en-us/nuget/reference/nuspec#dependency-groups

Dependency Groups are one example. Might be a good idea to take a quick look at the schema.

**Review by nkolev92 — COMMENTED** [review]
I think this looks a lot better, more scoped and just neater overall! 

I left a few comments, but don't wait on me obviously :)

File: src/NuGet.Core/NuGet.Packaging/Rules/InvalidUndottedFrameworkRule.cs (hunk 11 above)
Comment (zkat) [reply src/NuGet.Core/NuGet.Packaging/Rules/InvalidUndottedFrameworkRule.cs:204]: iirc, this is needed because other rules _do_ apply to normalized paths, but the paths we're operating on at this stage are still pre-normalization.

File: src/NuGet.Core/NuGet.Frameworks/NuGetFrameworkUtility.cs [hunk 14]
```diff
@@ -154,5 +154,21 @@ public static bool IsNetCore50AndUp(NuGetFramework framework)
                         framework.Framework,
                         FrameworkConstants.FrameworkIdentifiers.NetCore));
         }
+
+        public static bool FrameworkVersionHasRequiredDots(string frameworkString)
```
Comment (zivkan) [inline src/NuGet.Core/NuGet.Frameworks/NuGetFrameworkUtility.cs]: I'm not sure this is generally useful to customers using the NuGet SDK, so I'm not keen on it being part of the public API. It's only used in one place, NuGet.Packaging, so I'd suggest making it internal in there.

Also, the dot is recommended, rather than required. If it was required, we'd be failing restore.

File: src/NuGet.Core/NuGet.Packaging/GlobalSuppressions.cs [hunk 15]
```diff
@@ -309,3 +309,4 @@
...
 [assembly: SuppressMessage("ApiDesign", "RS0026:Do not add multiple public overloads with optional parameters", Justification = "API already shipped", Scope = "member", Target = "~M:NuGet.Packaging.PackageExtractor.InstallFromSourceAsync(NuGet.Packaging.Core.PackageIdentity,NuGet.Packaging.IPackageDownloader,NuGet.Packaging.VersionFolderPathResolver,NuGet.Packaging.PackageExtractionContext,System.Threading.CancellationToken,System.Guid)~System.Threading.Tasks.Task{System.Boolean}")]
 [assembly: SuppressMessage("ApiDesign", "RS0027:Public API with optional parameter(s) should have the most parameters amongst its public overloads.", Justification = "API already shipped", Scope = "member", Target = "~M:NuGet.Packaging.PackageArchiveReader.#ctor(System.String,NuGet.Frameworks.IFrameworkNameProvider,NuGet.Frameworks.IFrameworkCompatibilityProvider)")]
+[assembly: SuppressMessage("Globalization", "CA1307:Specify StringComparison", Justification = "These are not localized strings", Scope = "member", Target = "~M:NuGet.ContentModel.ContentItemCollection.GroupComparer.Equals(NuGet.ContentModel.ContentItem,NuGet.ContentModel.ContentItem)~System.Boolean")]
```
Comment (zivkan) [inline src/NuGet.Core/NuGet.Packaging/GlobalSuppressions.cs]: You can use `StringComparison.Ordinal` or `StringComparison.OrdinalIgnoreCase` to make it a culture independent comparison.

File: src/NuGet.Core/NuGet.Packaging/Rules/AnalysisResources.resx [hunk 16]
```diff
@@ -169,6 +169,10 @@
   <data name="InvalidPrereleaseDependencyWarning" xml:space="preserve">
     <value>A stable release of a package should not have a prerelease dependency. Either modify the version spec of dependency "{0}" or update the version field in the nuspec.</value>
   </data>
+  <data name="InvalidUndottedFrameworkWarning" xml:space="preserve">
+    <value>Packaged folders must include dots in their framework version numbers. This is required as of .NET 5.0. Please rename the following folder(s) to include dots as needed (e.g. 'net50' to 'net5.0'): {0}</value>
```
Comment (zivkan) [inline src/NuGet.Core/NuGet.Packaging/Rules/AnalysisResources.resx]: The first sentence in this message sounds a bit like 'net47' should be changed to 'net4.7'. It won't be in the list of folder names failing the rule, but to me that's what it sounds like.

Just an idea: What about, instead of doing string parsing to find if the framework version part of the TFM has a dot, we compare the input's string to `NuGetFramework.Parse(tfm).ToShortFolderName()`, then we can have an error message something along the lines of:
```text
Error NU5501: The following folders do not have the expected target framework folder name:
net50 (under lib/) should be net5.0
net50-windows7.0 (under build/) should be net5.0-windows7.0
```

I don't know if it's better to only report a folder once, like above, or for each file?
```
net50 (in build/net50/packageId.props) should be net5.0
net50 (in build/net50/packageId.targets) should be net5.0
net50 (in build/net50/myTasks.dll) should be net5.0
net50 (in build/net50/Newtonsoft.Json.dll) should be net5.0
```

We can still consider only .netcoreapp version >= 5.0, but perhaps this may minimise the potential for misunderstanding which tfm folders need to be renamed.

File: test/NuGet.Core.Tests/NuGet.Build.Tasks.Pack.Test/PackTaskLogicTests.cs [hunk 17]
```diff
@@ -64,6 +65,137 @@ public void PackTaskLogic_ProducesBasicPackage()
             }
         }
 
+        [Fact]
+        public void PackTaskLogic_WhenUsingCsproj_WarnsMissingDot()
```
Comment (zivkan) [inline test/NuGet.Core.Tests/NuGet.Build.Tasks.Pack.Test/PackTaskLogicTests.cs]: As you know, I don't like functional tests. Despite the fact that this project is under `NuGet.Core.Tests` instead of `NuGet.Core.FuncTests`, it:
* Runs all the code for a pack, from selecting files, generating the nuspec, generating the zip, and running all validation rules, not just the one you added in the production code.
* When it creates a zip file, it compresses the content, which is CPU invensive
* Any pack rules that read content from the package (several pack rules read the nuspec), need to decompress the file, which again wastes CPU time.

Given we already have func tests to ensure that validation rules are run on pack, I don't believe we need to, or should, run slow func tests to test individual pack validation rules.

Consider [`UpholdBuildConventionRule`](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Packaging/Rules/UpholdBuildConventionRule.cs). Its `public IEnumerable<PackagingLogMessage> Validate(PackageArchiveReader builder)` is a "thin" wrapper around two unit-testable `internal` methods, that are exhaustively tested in [`UpholdBuildConventionRuleTests.cs`](https://github.com/NuGet/NuGet.Client/blob/dev/test/NuGet.Core.Tests/NuGet.Packaging.Test/UpholdBuildConventionRuleTests.cs)

File: test/NuGet.Core.Tests/NuGet.Build.Tasks.Pack.Test/PackTaskLogicTests.cs [hunk 18]
```diff
@@ -64,6 +65,137 @@ public void PackTaskLogic_ProducesBasicPackage()
...
+                tc.Request.ContentTargetFolders = new string[] { "lib", "content", "contentFiles" };
+
+                // Act
+                tc.BuildPackage();
+
+                // Assert
+                var logger = (TestLogger)tc.Request.Logger;
+                var messages = logger.WarningMessages.ToArray();
+                Assert.True(messages[0].Contains("lib/net50/def.dll"));
+                Assert.True(messages[0].Contains("include dot"));
```
Comment (zivkan) [inline test/NuGet.Core.Tests/NuGet.Build.Tasks.Pack.Test/PackTaskLogicTests.cs]: Isn't NU5501 reported? We should check for that, rather than a short string, which may accidentally be used by some other validation rule.

File: test/NuGet.Core.Tests/NuGet.Build.Tasks.Pack.Test/PackTaskLogicTests.cs [hunk 19]
```diff
@@ -64,6 +65,137 @@ public void PackTaskLogic_ProducesBasicPackage()
...
+    <title>bartitle</title>
+    <authors>kat</authors>
+    <requireLicenseAcceptance>true</requireLicenseAcceptance>
+    <license type=""expression"">MIT</license>
+    <description>desc</description>
+    <releaseNotes>release notes</releaseNotes>
+    <copyright>msft</copyright>
+    <tags>foo bar</tags>
+    <dependencies>
+        <group targetFramework=""net50"">
```
Comment (zivkan) [inline test/NuGet.Core.Tests/NuGet.Build.Tasks.Pack.Test/PackTaskLogicTests.cs]: This test uses `net50` in both this dependency group, and also in `lib/net50/a.dll`, but the assertions don't validate that both `net50` usages get reported.

Therefore, the test isn't validating what some people reading the test will think that it is validating.

File: test/NuGet.Core.Tests/NuGet.Build.Tasks.Pack.Test/PackTaskLogicTests.cs [hunk 20]
```diff
@@ -64,6 +65,137 @@ public void PackTaskLogic_ProducesBasicPackage()
...
+                // Assert
+                var logger = (TestLogger)tc.Request.Logger;
+                var messages = logger.WarningMessages.ToArray();
+                Assert.True(messages[0].Contains("lib/net50/def.dll"));
+                Assert.True(messages[0].Contains("include dot"));
+            }
+        }
+
+        [Fact]
+        public void PackTaskLogic_WarnsMissingDot_UsingNuspec()
```
Comment (zivkan) [inline test/NuGet.Core.Tests/NuGet.Build.Tasks.Pack.Test/PackTaskLogicTests.cs]: I dodn't look at the other tests in this file, but the general naming pattern some people have been pushing for recently is: `<MethodBeingTested>_<Scenario>_<ExpectedResults>`.  It's not so clear what `MethodBeingTested` should be for functional tests, but the expected result should be at the end of the name

File: test/NuGet.Core.Tests/NuGet.Build.Tasks.Pack.Test/PackTaskLogicTests.cs [hunk 21]
```diff
@@ -64,6 +65,137 @@ public void PackTaskLogic_ProducesBasicPackage()
...
+    <title>bartitle</title>
+    <authors>kat</authors>
+    <requireLicenseAcceptance>true</requireLicenseAcceptance>
+    <license type=""expression"">MIT</license>
+    <description>desc</description>
+    <releaseNotes>release notes</releaseNotes>
+    <copyright>msft</copyright>
+    <tags>foo bar</tags>
+    <dependencies>
+        <group targetFramework=""net50-windows7.0"">
```
Comment (zivkan) [inline test/NuGet.Core.Tests/NuGet.Build.Tasks.Pack.Test/PackTaskLogicTests.cs]: like the other test, it's not clear if this test is supposed to be testing TFMs in the nuspec or not. Historically it's been valid for a nupkg to not have any dependencies. As of last year, it will generate a warning, so your assertions can take that into account and ignore that warning, and check only for NU5501. That makes it more clear, in my opinion, what this test is actually testing.

File: src/NuGet.Core/NuGet.Packaging/PublicAPI/net472/PublicAPI.Unshipped.txt [hunk 22]
```diff
@@ -1 +1,5 @@
+static NuGet.Packaging.Rules.AnalysisResources.InvalidUndottedFrameworkInDependencyGroupsWarning.get -> string
```
Comment (zivkan) [inline src/NuGet.Core/NuGet.Packaging/PublicAPI/net472/PublicAPI.Unshipped.txt:1]: scope creep:  It's just a guess, but I guess that `AnalysisResources` are public only so tests can access the message text without using `InternalsVisibleTo`. These strings are not useful to customers using the SDK, because we could change the `{x}` placeholders at any time.

What do you think about making a new resx file that's internal, so we stop getting these public API changes just because we're adding messages?

File: src/NuGet.Core/NuGet.Packaging/Rules/InvalidUndottedFrameworkRule.cs [hunk 23]
```diff
@@ -17,14 +19,167 @@ namespace NuGet.Packaging.Rules
...
         {
+            // IPackageRule expects to get messages passed in, but we output
+            // different kinds of messages, so each individual validator is
+            // responsible for its own messaging.
             MessageFormat = messageFormat;
         }
 
         public IEnumerable<PackagingLogMessage> Validate(PackageArchiveReader builder)
+        {
+            _xml = LoadXml(builder.GetNuspec());
```
Comment (zivkan) [inline src/NuGet.Core/NuGet.Packaging/Rules/InvalidUndottedFrameworkRule.cs]: ~don't we have a "strongly typed" nuspec model, so you don't have to do your own XML parsing? Genuine question, I have no idea, I just assumed we would.~

We do, `NuspecReader`. However, this validation rule needs the unparsed framework strings. Can you please add a comment explaining why `NuspecReader` wasn't used.

Comment (zivkan) [inline src/NuGet.Core/NuGet.Packaging/Rules/InvalidUndottedFrameworkRule.cs]: If each of the `ValidateXxxx` methods took in an input parameter, and were `internal static`, rather than `private`, they'd be unit testable, making it much easier and faster to write exhaustive tests for all edge cases.

File: src/NuGet.Core/NuGet.Packaging/PublicAPI/net472/PublicAPI.Unshipped.txt (hunk 22 above)
Comment (zkat) [reply src/NuGet.Core/NuGet.Packaging/PublicAPI/net472/PublicAPI.Unshipped.txt:1]: I'm gonna treat this as scope creep. Maybe we can create an issue to find a better solution for AnalysisResources in general?

**Comment by zkat** [comment]
Note: Tests are green, but GH apparently didn't update properly.

**Comment by zivkan** [comment]
> Note: Tests are green, but GH apparently didn't update properly.

We don't use CI integration with github, because .... reasons. We call GitHub's REST API from our own build scripts. Our builds recently to work a little bit differently, and now the "update github status" script doesn't correctly handle windows unit tests failing :(

This means if unit tests fail, we have to spin a new build until windows unit tests are green. Any other failure can be retried, but windws unit test failures will "break" github check status. I've already created a CI initiative task.

File: src/NuGet.Core/NuGet.Packaging/Rules/InvalidUndottedFrameworkRule.cs [hunk 24]
```diff
@@ -0,0 +1,295 @@
...
+        private static readonly char[] CommaArray = new char[] { ',' };
+
+        public string MessageFormat { get; }
+
+        public InvalidUndottedFrameworkRule(string messageFormat)
+        {
+            // IPackageRule expects to get messages passed in, but we output
+            // different kinds of messages, so each individual validator is
+            // responsible for its own messaging.
+            MessageFormat = messageFormat;
```
Comment (zivkan) [inline src/NuGet.Core/NuGet.Packaging/Rules/InvalidUndottedFrameworkRule.cs]: Why not remove the parameter from the constructor?

[There are already 3 rules that use the default constructor](https://github.com/NuGet/NuGet.Client/blob/16be25216b699f48d5e4d0baed86a62c78acadbe/src/NuGet.Core/NuGet.Packaging/Rules/RuleSet.cs#L29-L31)

File: test/NuGet.Core.Tests/NuGet.Build.Tasks.Pack.Test/PackTaskLogicTests.cs [hunk 25]
```diff
@@ -9,6 +9,7 @@
 using NuGet.Commands;
 using NuGet.Frameworks;
 using NuGet.Packaging;
+using NuGet.Packaging.Core;
```
Comment (zivkan) [inline test/NuGet.Core.Tests/NuGet.Build.Tasks.Pack.Test/PackTaskLogicTests.cs]: this is unnecessary, given there are no other changes in the file

File: test/NuGet.Core.Tests/NuGet.Frameworks.Test/NuGetFrameworkUtilityTests.cs [hunk 26]
```diff
@@ -117,5 +117,6 @@ public void NuGetFrameworkUtility_GetNearest_Duplicates()
             Assert.Null(nearestWithSelector);
             Assert.Null(nearest);
         }
+
```
Comment (zivkan) [inline test/NuGet.Core.Tests/NuGet.Frameworks.Test/NuGetFrameworkUtilityTests.cs]: there are no other changes in this file. Ideally it's reverted.

File: test/NuGet.Core.Tests/NuGet.Packaging.Test/InvalidUndottedFrameworkRuleTests.cs [hunk 27]
```diff
@@ -0,0 +1,124 @@
...
+using System.Xml.Linq;
+using NuGet.Commands;
+using NuGet.Common;
+using NuGet.Packaging.Rules;
+using NuGet.Test.Utility;
+using Xunit;
+
+namespace NuGet.Packaging.Test
+{
+    public class InvalidUndottedFrameworkRuleTests
```
Comment (zivkan) [inline test/NuGet.Core.Tests/NuGet.Packaging.Test/InvalidUndottedFrameworkRuleTests.cs:18]: Apart from the `FrameworkVersionHasDesiredDots` test, the others are missing tests to ensure they don't generate the warning when:
* The package correctly dots `net5.0`
* The package doesn't contain any TFM that should have dots (for example, contains only netfx, netstandard or netcoreapp tfms)

I can't think of any other missing test cases, but particularly when you have easy to write unit tests, please remember to have tests for: happy case, error case, edge cases.


Another missing thing from the tests are validating that they generate the correct number of warnings. If any of the XML elements has both `net50` and `net50-windows7.0`, will it generate the correct number of warnings? Taking a look at the code, it appears that you'll be generating one warning per TFM, whereas I thought that an earlier comment requested that we generate one message per "message format", with a list of all the TFMs failing the validation.

Finally, and this one is much more subjective, but particularly when generating messages with multiple generated parts, I think it might be useful to have a test to ensure the generated message looks correct: https://github.com/NuGet/NuGet.Client/blob/16be25216b699f48d5e4d0baed86a62c78acadbe/test/NuGet.Core.Tests/NuGet.Packaging.Test/UpholdBuildConventionRuleTests.cs#L139-L143

**Review by zivkan — APPROVED** [review]
I know there are time pressures to get this in, so it's acceptable to me that it gets merged as-is. But particularly my test comments should be considered for normal developments. Maybe my other comments are a bit nit-picky

File: test/NuGet.Core.Tests/NuGet.Packaging.Test/InvalidUndottedFrameworkRuleTests.cs (hunk 27 above)
Comment (zkat) [reply test/NuGet.Core.Tests/NuGet.Packaging.Test/InvalidUndottedFrameworkRuleTests.cs:18]: > Taking a look at the code, it appears that you'll be generating one warning per TFM, whereas I thought that an earlier comment requested that we generate one message per "message format", with a list of all the TFMs failing the validation.

The code actually generates one message per "check type", each with all relevant TFMed values concatenated together into a single message. Additionally, a general "oops you got a NU5501" message is appended if any of these generate a warning.

Re: tests, working on those now, should be ready soon.

Comment (zkat) [reply test/NuGet.Core.Tests/NuGet.Packaging.Test/InvalidUndottedFrameworkRuleTests.cs:18]: I've omitted the tests that check specific messages, and that the right number of messages is displayed, but I've also done a small refactor so that even more stuff is unit-testable, so this can be done later, after the current rush.
